File: parent/45 -- Jump Game II/tutorial_dp.py
class Solution:
    def jump(self, nums: list[int]) -> int:
        # Bottom-up DP
        # Time: O(n^2)
        # Space: O(n)
        
        n = len(nums)
        
        # Initialize DP array with infinity. 
        # dp[i] will store the min jumps to reach index i.
        dp = [float('inf')] * n
        dp[0] = 0  # 0 jumps needed to stand at the starting line
        
        for i in range(n):
            # From the current position i, update all reachable future positions
            for step in range(1, nums[i] + 1):
                if i + step < n:
                    # The min jumps to reach (i + step) is either its current known 
                    # min jumps, or the jumps to reach i plus 1 more jump.
                    dp[i + step] = min(dp[i + step], dp[i] + 1)
                    
        return dp[-1]

File: parent/45 -- Jump Game II/tutorial_dp.go
package main

import "math"

func jump(nums []int) int {
	n := len(nums)
	dp := make([]int, n)
	for i := range n {
		dp[i] = math.MaxInt
	}

	dp[0] = 0

	for i := range n {
		for step := 1; step <= nums[i]; step++ {
			if i+step < n {
				dp[i+step] = min(dp[i+step], dp[i]+1)
			}
		}
	}

	return dp[len(dp)-1]
}
